### PR TITLE
[WIP] Collections UI

### PR DESCRIPTION
**Pull request by tsop14**
NB The button to reveal the collections panel is currently not being shown. 

styling

⚡️ Adds a collections panel on the LHS
⚡️ Creates a folder structure with JSON from collections api
⚡️ Folders can show/hide their children
⚡️ Leaves can be deleted
⚡️ Child folders can be created on any node
⚡️ Initial styling on panel

TODO
- [ ] add image to collection when dropped on list
- [ ]  turn off click completion when drag icon dropped on its original image (currently takes you to the image page - see end of video)
- [ ] undo button / cancel a mistaken drop
- [ ] create a search term for collections
- [ ] return collection when clicking on folder
- [ ] sort images in collection view by date added to the collection 
- [ ] autofocus on new collection's text input

### CODE REVIEW

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.css [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    display: block;
+}
+
+.node__info:hover .node__action {
+    display: block;
+}
+
+.node__marker {
+    position: absolute;
+    background: hotpink;
```
Comment (akash1810) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.css]: Nice! Does it fit the Grid's theme though?

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+grCollectionsPanel.directive('grNode', ['$parse', '$compile', function($parse, $compile) {
+    return {
+        restrict: 'E',
+        scope: {
+            node: '=grNode',
+            nodeList: '=grNodeList'
+        },
+        template: `
+        <div ng:init="node = ctrl.node; showChildren = true;">
```
Comment (akash1810) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js]: Can we pull the template into its own file so it can be linted?

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+            <div class="node__children"></div>
+        </div>
+        `,
+        controller: 'GrNodeCtrl',
+        controllerAs: 'ctrl',
+        bindToController: true,
+        link: function(scope, element, attrs, ctrl) {
+            if (ctrl.node.data.children.length > 0) {
+                $compile(`<gr-nodes class="tree" ng:show="showChildren" gr:nodes="ctrl.node.data.children"></gr-nodesclass>`)(scope, cloned => {
```
Comment (akash1810) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js]: `</gr-nodesclass>` = `</gr-nodes>` ?

**Comment by alastairjardine** [comment]
@tsop14 @jamesgorrie Are the collections addressable? e.g. http://media.gutools.co.uk/collection/arts/films/skyfall

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.css (hunk 1 above)
Comment (alastairjardine) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.css]: @akash1810 Few minor tweaks coming here ;)

**Comment by tsop14** [comment]
> Are the collections addressable? 

Not yet. I believe James is working on this.

**Comment by jamesgorrie** [comment]
@alastairjardine are you asking about what the URL might look like if you're looking to see all the results? Probably consistent with the rest: `/search?collection=king/kang/kong` - good to chat about why? But probably not on this PR.

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.css (hunk 1 above)
Comment (tsop14) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.css]: > Nice! Does it fit the Grid's theme though?

Although I'm pro hotpink, I think not. @alastairjardine: the blue used elsewhere e.g. 'upload' button?

Comment (jamesgorrie) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.css]: Looks like iplayer. We have `peachpuff`, `saddlebrown` or `goldenrod` :boom:

**Comment by alastairjardine** [comment]
@jamesgorrie Just wanted to ensure it's usable, that's all! (i.e. not a random string)

**Comment by jamesgorrie** [comment]
@alastairjardine the collections have a `pathId` - [it's described here](https://github.com/guardian/grid/blob/master/common-lib/src/test/scala/com/gu/mediaservice/lib/collections/CollectionsManagerTest.scala#L10-L24) - let me know if that doesn't make sense.

**Comment by alastairjardine** [comment]
@jamesgorrie Looks good, though I can foresee lots of `%20` misery though :,(

**Comment by jamesgorrie** [comment]
>  I can foresee lots of %20 misery though 

Not in Firefox. That's just a limitation of URLs unfortunately, but only on read, if you type it in the URL, you're good to go. That's just URL's unfortunately.

**Comment by alastairjardine** [comment]
> Unless you want to open a whole can of works and presume you know what you should substitute spaces with.

`-`?

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js (hunk 2 above)
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js]: We do template linting?

**Comment by jamesgorrie** [comment]
:+1: to @tsop14's contribution - I'll need one on mine...

File: kahuna/public/js/services/api/collections-api.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        return collection.perform('delete');
+    }
+
+    function addCollection(newCollectionPath) {
+        return mediaApi.root.follow('collections').post({data: newCollectionPath});
+    }
+
+    function addChildTo(node, childName) {
+        return node.perform('add-child', {body: {data: childName}}).then(childResource => {
+            const updatedChildren = node.data.children = [childResource].concat(node.data.children);
```
Comment (theefer) [inline kahuna/public/js/services/api/collections-api.js]: Minor detail: technically, the child will be inserted in the child list using alphabetical sorting, rather than always appending? I don't think it's worth worrying about, but just wanted to point out the potential discrepancy.

File: kahuna/public/js/services/api/collections-api.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            return updatedChildren;
+        });
+    }
+
+    function isDeletable(node) {
+        return node.getAction('remove').then(d => angular.isDefined(d));
```
Comment (theefer) [inline kahuna/public/js/services/api/collections-api.js]: Note: I think you could even simplify to `.then(angular.isDefined)` if you preferred, but I don't mind either way.

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel-node.html [hunk 6]
```diff
@@ -0,0 +1,48 @@
+<div ng:init="node = ctrl.node; showChildren = true;">
+    <div class="node__info flex-container" gr:add-to-collection>
+        <div class="node__marker"></div>
+
+        <div class="node__spacer" ng:if="node.data.children.length === 0">
+            <gr-icon></gr-icon>
```
Comment (theefer) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel-node.html]: What's this?

Comment (jamesgorrie) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel-node.html]: It's so we don't have to recreate the styling for the icon to create the exact amount of space it takes up.

File: kahuna/public/js/search/results.html [hunk 7]
```diff
@@ -15,7 +15,22 @@
         <div class="results-toolbar"
             ng:class="{'results-toolbar--selection-mode': ctrl.metadataPanelAvailable}">
 
-            <div class="results-toolbar-item results-toolbar-item--static results-toolbar-item--left
+            <div ng-show="false"
+                class="results-toolbar-item
+                results-toolbar-item--left
+                clickable"
+                 ng:click="ctrl.toggleCollectionsPanel()">
```
Comment (theefer) [inline kahuna/public/js/search/results.html]: If an element can be clicked, please use a clickable control for accessibility reasons and making things more obvious. Here `button` would make most sense.

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    $rootScope.$on(
+        `ui:panels:${panelName}:updated`,
+        () => {
+            ctrl.isVisible = panelService.isVisible(panelName);
+        }
+    );
+
+    collections.getCollections().then(collections => {
+        ctrl.collections = collections.data.children;
+    });
```
Comment (theefer) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js]: Not urgent, but we should really add error handler in such places to display an error message ("oops, collections not available, please try again later") in case the collections service is not available.

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        controller: 'GrNodeCtrl',
+        controllerAs: 'ctrl',
+        bindToController: true,
+        link: function(scope, element, attrs, ctrl) {
+            if (ctrl.node.data.children.length > 0) {
+                $compile(`<gr-nodes class="tree"
+                ng:show="showChildren"
+                gr:nodes="ctrl.node.data.children"></gr-nodes>`)(scope, cloned => {
+                    element.find('.node__children').append(cloned);
+                });
```
Comment (theefer) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js]: Why the inline `$compile` here?

**Comment by theefer** [comment]
A few comments above, but it's looking great!

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js (hunk 8 above)
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js]: Definitely - I think this service needs a lot of work - it's just a works-enough to show people. I will add some TODOs

File: kahuna/public/js/search/results.html (hunk 7 above)
Comment (jamesgorrie) [inline kahuna/public/js/search/results.html]: Done

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js (hunk 9 above)
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js]: Otherwise we land up in a massive angular compile loop.
TL;DR; If this bit of the template is included in the parent, angular tries to compile the template, pulls in itself, tries to compile the template, includes itself - basically templates don't deal well with recursion.

There's a few solutions, this seemed to make the most sense.

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.css (hunk 1 above)
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.css]: Back to blue.

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js (hunk 3 above)
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js]: Nice cathc, IntelliJ autocompletion.

File: kahuna/public/js/services/api/collections-api.js (hunk 4 above)
Comment (jamesgorrie) [inline kahuna/public/js/services/api/collections-api.js]: Yea - we know that it will be at the top of the list after inserting, but not on refresh - we thought this the lesser of appending it, there is also the potential  I would like to look at of potentially returning the whole tree again from the server, but this was quicker.

File: kahuna/public/js/services/api/collections-api.js (hunk 5 above)
Comment (jamesgorrie) [inline kahuna/public/js/services/api/collections-api.js]: Done.

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js (hunk 8 above)
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js]: Added some basic error handling - there's a wider problem with contextual error handling as we need a way to stop it being caught by our [global handler](https://github.com/guardian/grid/blob/master/kahuna/public/js/main.js#L174-L182).

Potentially decide to catch all errors contextually.

**Comment by jamesgorrie** [comment]
@theefer tidied up a few lose ends.

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.css [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+.node__name {
+    padding: 10px;
+}
+
+.node__toggler,
+.node__spacer,
+.node__action {
+    padding-left: 10px;
+    padding-right: 10px;
+    z-index: 100;
```
Comment (akash1810) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.css]: :(

**Comment by akash1810** [comment]
looks good!

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.css (hunk 10 above)
Comment (jamesgorrie) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.css]: Changed it to 2.

**Comment by jamesgorrie** [comment]
Update on Gif
collections

File: kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js (hunk 9 above)
Comment (theefer) [inline kahuna/public/js/components/gr-collections-panel/gr-collections-panel.js]: Interesting, thanks!

**Comment by theefer** [comment]
:+1:
